presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayAllocator.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.api.Experimental;
+
+/**
+ * Manages the creation and return of primitive arrays, to be used within an operator to avoid repeated array allocation.
+ * Typically this will be used within {@link BlockFlattener}.
+ */
+@Experimental
+public interface ArrayAllocator
+{
+    int[] borrowIntArray(int positionCount);
+
+    void returnArray(int[] array);
+
+    int getOutstandingArrays();
```

Comment:
`borrowIntArray` suggests that you expect more methods like this for different types of arrays in the future, e.g. borrowDoubleArray, borrowBooleanArray, etc. If that's the case, perhaps, getOutstandingArrays should be renamed to getOutstandingIntArrays? (Would it also make sense to replace outstanding with borrowed?)

Reply:
I replaced `Outstanding` with `Borrowed`.  I don't know at this time if it makes sense in the future to separate the counts of the different types of arrays borrowed, but this API is subject to change and we can do it if or when it becomes necessary.

Reply:
Added a comment describing the intent of it